msrestazure/azure_active_directory.py

```diff
@@ -131,61 +76,108 @@ def _configure(self, **kwargs):
         self.cred_store = kwargs.get('keyring', self._keyring)
         self.resource = kwargs.get('resource', resource)
-        self.proxies = kwargs.get('proxies')
-        self.timeout = kwargs.get('timeout')
-        self.state = oauth.oauth2_session.generate_token()
+        self._proxies = kwargs.get('proxies')
+        self._timeout = kwargs.get('timeout')
+        self._cache = kwargs.get('cache')
         self.store_key = "{}_{}".format(
             auth_endpoint.strip('/'), self.store_key)
+        self.secret = None
+        self._context = None  # Future ADAL context
+
+    def _create_adal_context(self):
+        authority_url = self.cloud_environment.endpoints.active_directory
+        is_adfs = bool(re.match('.+(/adfs|/adfs/)$', authority_url, re.I))
+        if is_adfs:
+            authority_url = authority_url.rstrip('/')  # workaround: ADAL is known to reject auth urls with trailing /
+        else:
+            authority_url = authority_url + '/' + self._tenant
+
+        self._context = adal.AuthenticationContext(
+            authority_url,
+            timeout=self._timeout,
+            verify_ssl=self._verify,
+            proxies=self._proxies,
+            validate_authority=not is_adfs,
+            cache=self._cache,
+            api_version=None
+        )
 
-    def _check_state(self, response):
-        """Validate state returned by AAD server.
+    def _destroy_adal_context(self):
+        self._context = None
 
-        :param str response: URL returned by server redirect.
-        :raises: ValueError if state does not match that of the request.
-        :rtype: None
-        """
-        query = parse_qs(urlparse(response).query)
-        if self.state not in query.get('state', []):
-            raise ValueError(
-                "State received from server does not match that of request.")
+    @property
+    def verify(self):
+        return self._verify
+
+    @verify.setter
+    def verify(self, value):
+        self._verify = value
+        self._destroy_adal_context()
+
+    @property
+    def proxies(self):
+        return self._proxies
+
+    @proxies.setter
+    def proxies(self, value):
+        self._proxies = value
+        self._destroy_adal_context()
+
+    @property
+    def timeout(self):
+        return self._timeout
+
+    @timeout.setter
+    def timeout(self, value):
+        self._timeout = value
+        self._destroy_adal_context()
+
+    @property
+    def cloud_environment(self):
+        return self._cloud_environment
+
+    @cloud_environment.setter
+    def cloud_environment(self, value):
+        self._cloud_environment = value
+        self._destroy_adal_context()
 
     def _convert_token(self, token):
         """Convert token fields from camel case.
 
         :param dict token: An authentication token.
         :rtype: dict
         """
+        # Beware that ADAL returns a pointer to its own dict, do
+        # NOT change it in place
+        token = token.copy()
+
```

Comment:
I think, I'm running into this bug with the following code. Can this change be hotfixed on the latest release? It appears this PR may take some time.

```
AUTHORITY_URL = "https://login.microsoftonline.com" + '/' + os.environ["TENANT"]
auth_context = adal.AuthenticationContext(AUTHORITY_URL, api_version=None)
resource = "https://vault.azure.net"
client_id = os.environ["CLIENT_ID"]
client_secret = os.environ["CLIENT_SECRET"]
token = auth_context.acquire_token_with_client_credentials(resource, client_id, client_secret)
credentials = AADTokenCredentials(token, client_id)
print("Got one token. Trying to get the token again")
token = auth_context.acquire_token_with_client_credentials(resource, client_id, client_secret)
credentials = AADTokenCredentials(token, client_id)
print("Got one token second time as well")
```
The second print never gets printed and I get this exception.
```
  File "/Users/prburgu/Downloads/test1/lib/python2.7/site-packages/dateutil/parser/_parser.py", line 76, in __init__
    '{itype}'.format(itype=instream.__class__.__name__))
TypeError: Parser must be a string or character stream, not float
```

Reply:
Hi @prburgu 
Sounds plausible, but could I get the complete stacktrace please? Not just the last line.

Also, this PR will be merged as 0.5.0, but this creates some conflicts right now with some packages that requires 0.4.x. I hope all will be merged and released by this summer.

Reply:
Actually, I've worked around this issue in my code, so there is no urgent need for hotfix and I'll leave it up to you on that. Here is the complete stacktrace. 

```
Traceback (most recent call last):
  File "adal_test.py", line 13, in <module>
    token = auth_context.acquire_token_with_client_credentials(resource, client_id, client_secret)
  File "/Users/prburgu/Downloads/test1/lib/python2.7/site-packages/adal/authentication_context.py", line 179, in acquire_token_with_client_credentials
    return self._acquire_token(token_func)
  File "/Users/prburgu/Downloads/test1/lib/python2.7/site-packages/adal/authentication_context.py", line 128, in _acquire_token
    return token_func(self)
  File "/Users/prburgu/Downloads/test1/lib/python2.7/site-packages/adal/authentication_context.py", line 177, in token_func
    return token_request.get_token_with_client_credentials(client_secret)
  File "/Users/prburgu/Downloads/test1/lib/python2.7/site-packages/adal/token_request.py", line 306, in get_token_with_client_credentials
    token = self._find_token_from_cache()
  File "/Users/prburgu/Downloads/test1/lib/python2.7/site-packages/adal/token_request.py", line 128, in _find_token_from_cache
    return self._cache_driver.find(cache_query)
  File "/Users/prburgu/Downloads/test1/lib/python2.7/site-packages/adal/cache_driver.py", line 201, in find
    is_resource_tenant_specific)
  File "/Users/prburgu/Downloads/test1/lib/python2.7/site-packages/adal/cache_driver.py", line 167, in _refresh_entry_if_necessary
    expiry_date = parser.parse(entry[TokenResponseFields.EXPIRES_ON])
  File "/Users/prburgu/Downloads/test1/lib/python2.7/site-packages/dateutil/parser/_parser.py", line 1356, in parse
    return DEFAULTPARSER.parse(timestr, **kwargs)
  File "/Users/prburgu/Downloads/test1/lib/python2.7/site-packages/dateutil/parser/_parser.py", line 645, in parse
    res, skipped_tokens = self._parse(timestr, **kwargs)
  File "/Users/prburgu/Downloads/test1/lib/python2.7/site-packages/dateutil/parser/_parser.py", line 721, in _parse
    l = _timelex.split(timestr)         # Splits the timestr into tokens
  File "/Users/prburgu/Downloads/test1/lib/python2.7/site-packages/dateutil/parser/_parser.py", line 207, in split
    return list(cls(s))
  File "/Users/prburgu/Downloads/test1/lib/python2.7/site-packages/dateutil/parser/_parser.py", line 76, in __init__
    '{itype}'.format(itype=instream.__class__.__name__))
TypeError: Parser must be a string or character stream, not float
```

Reply:
I was looking for the `_find_token_from_cache` and it's there :). You're right, that's the correct fix.
I'd rather wait for this PR to be merged if it's ok for you.